module/zfs/vdev_raidz_math.c

```diff
@@ -0,0 +1,713 @@
+	const size_t op_name_max = (KSTAT_STRLEN-2) > impl_name_len ?
+		KSTAT_STRLEN - impl_name_len - 2 : 0;
+
+	for (i = 0; i < RAIDZ_GEN_NUM; i++) {
+		strncpy(rs->gen_kstat[i].name, name, impl_name_len);
+		strncpy(rs->gen_kstat[i].name + impl_name_len, "_", 1);
+		strncpy(rs->gen_kstat[i].name + impl_name_len + 1,
+			raidz_gen_name[i], op_name_max);
+
+		rs->gen_kstat[i].data_type = KSTAT_DATA_UINT64;
+		rs->gen_kstat[i].value.ui64  = 0;
+	}
+
+	for (i = 0; i < RAIDZ_REC_NUM; i++) {
+		strncpy(rs->rec_kstat[i].name, name, impl_name_len);
+		strncpy(rs->rec_kstat[i].name + impl_name_len, "_", 1);
+		strncpy(rs->rec_kstat[i].name + impl_name_len + 1,
+			raidz_rec_name[i], op_name_max);
+
+		rs->rec_kstat[i].data_type = KSTAT_DATA_UINT64;
+		rs->rec_kstat[i].value.ui64  = 0;
+	}
+}
+
+
+/* Number of pages for benchmark: 2^BENCH_SIZE_ORD */
+#define	BENCH_D_COLS	(8)
+#define	BENCH_COLS	(3 + BENCH_D_COLS)
+#define	BENCH_COL_SIZE	(512)
+#define	BENCH_ZIO_SIZE	(BENCH_COL_SIZE * BENCH_D_COLS)
+#define	BENCH_NS	(1ULL*1000ULL*1000ULL) /* 50ms */
+
+#define	BENCH_RES_FMT	"RAIDZ_math: [%-7s] %-8s %5ld.%.3ld MB/s\n"
+#define	BENCH_SKIP_FMT	"RAIDZ_math: [%-7s] is not supported.\n"
+
+#if defined(_KERNEL)
+#define	bench_print(x...)	cmn_err(CE_CONT, x)
+#else
+#define	bench_print(x...)	do { } while (0)
+#endif
+
+void
+vdev_raidz_math_init(void)
+{
+	const int impl_num = ARRAY_SIZE(raidz_all_maths);
+	const int tgtidx[3] = {3, 4, 5};
+	raidz_math_ops_t const *curr_impl;
+	unsigned long long run_count, best_run_count, speed;
+	hrtime_t t_start;
+	zio_t *bench_zio = NULL;
+	raidz_map_t *bench_rm = NULL;
+	uint64_t bench_parity;
+	int i, c, fn, impl;
+	int bench_rec_orig[7][3] = {
+		{1, 2, 3}, {0, 2, 3},
+		{0, 1, 3}, {2, 3, 4},
+		{1, 3, 4}, {0, 3, 4},
+		{3, 4, 5}
+	};
+
+	/* init & vdev_raidz_impl_lock */
+	rw_init(&vdev_raidz_impl_lock, NULL, RW_DEFAULT, NULL);
+
+#if !defined(_KERNEL)
+	/*
+	 * It is possible that this routine is called multiple times
+	 * inside user-space test. Check if we already initialized the
+	 * fastest ops and skip the benchmark.
+	 */
+	if (raidz_math_initialized) {
+		goto init_done;
+	}
+#endif
+
+	/* init scalar multiplication lookup table */
+	raidz_init_scalar_mul_lt();
+
+	/* move supported maths into raidz_supp_maths */
+	for (i = 0, c = 0; i < impl_num; i++) {
+		curr_impl = raidz_all_maths[i];
+
+		/* init kstat */
+		init_raidz_kstat(raidz_math_kstats + i, curr_impl->name);
+
+		if (curr_impl->is_supported()) {
+			raidz_supp_maths[c++] = (raidz_math_ops_t *) curr_impl;
+		} else {
+			bench_print(BENCH_SKIP_FMT, curr_impl->name);
+			continue;
+		}
+	}
+	raidz_supp_maths_cnt = c;	/* number of supported algos */
+
+	/* init  kstat for original algo  routines */
+	init_raidz_kstat(&(raidz_math_kstats[impl_num]), "original");
+
+	/* Fake an zio and run the benchmark on it */
+	bench_zio = kmem_alloc(sizeof (zio_t), KM_SLEEP);
+	VERIFY(bench_zio);
+	bzero(bench_zio, sizeof (zio_t));
```

Comment:
Ick, I hadn't noticed that.  Well then let's just tweak it to use `kmem_zalloc()` which will let you remove the VERIFY since the KM_SLEEP allocation can't fail and the `bzero()` since `kmem_zalloc()` will zero-fill.